### ISSUE TITLE
Allow for mime.types list

### ISSUE THREAD

**Issue by DavidGriffith** [gophernicus/gophernicus#3]
Would someone please add support for a mime.types file?  I maintain a gopher mirror of the Interactive Fiction Archive, and a feature like this would be much appreciated because of the wide variety of odd filename extensions.

**Comment by hb9kns** [comment]
Can you please give an example of the format(s) you would like to have supported?

**Comment by DavidGriffith** [comment]
Very simply, this: ````<type> <extension>````.  I maintain a gopher mirror of the Interactive Fiction Archive at [gopher://gopher.661.org/](url).  There are a lot of odd filename extensions in there that are rarely found elsewhere that often get served up as text rather than octet-stream.

````
application/octet-stream LHA
application/octet-stream lha
application/octet-stream SHK
application/octet-stream shk
application/octet-stream SDK
application/octet-stream sdk
````

**Comment by fosslinux** [comment]
So, you want a file names `mime.types`, with the syntax you specified above, so that when you put a link to a file, it uses that mime type instead of text? In `README.Gophermap`, it states:

     M MIME file (mbox, emails etc)

would this do the trick?

**Comment by DavidGriffith** [comment]
That filetype character is for marking files containing email.  It doesn't affect how the daemon treats unfamiliar filename extensions.

**Comment by hb9kns** [comment]
Wouldn't it make more sense to use a file containing e.g

    bmp I bitmap
    png I portable network graphics
    doc d Word
    docx d "new" Word (pseudo-open)
    odf d OpenDocumentFormat
    zip 5 zipfile
    tar 5 tape archive
    tgz 5 gzipped tape archive
    lha 5 some old archive I don't know
    mp3 s MP3
    sh 0 shell script

etc, with extension-gophertype-comment lines?
Mime types don't make much sense neither for gopher server nor clients, methinks.

**Comment by DavidGriffith** [comment]
I don't follow what you're suggesting.

**Comment by hb9kns** [comment]
Sorry, maybe bad wording on my side. ^-^
As I don't see the real use of mime types in gopher (neither for server nor client), because there is nothing indicating to the server what type a file might have (except if you have some meta file on your system for each published file) or to communicate this to the client (except if you extend the protocol), we have to rely on file extension and gopher file type.
But then we just need a file explaining to the server "extension X is gopher file type Y" and we might add an explanation/comment "Z" of this in the gophermap. The columns X/Y/Z are what I wrote in that example above.

Or maybe *I completely don't get* what you're trying to say/suggest..?

**Comment by hb9kns** [comment]
Gophernicus already has a number of predefined suffix-gophertype pairs in `gophernicus.h.` In addition, one can optionally give more pairs through the command line, e.g with `-e md=0` resulting in `*.md` returned as text file (type 0).

So you could easily add `"lha", "9", "shk", "9", "sdk", "9"` to `FILETYPES` in `gophernicus.h` or give options `-e lha=9 -e shk=9 -e sdk=9` on the command line to achieve what you suggested with your mime-type example file.

Therefore, and because I still don't see where that mime-type information would come from, I think we should close this issue.

**Comment by DavidGriffith** [comment]
Here's the mime.conf file from the IF Archive:

````
# MIME types list for the IF Archive (and other Apache configuration).

# All our HTML files are UTF-8.
AddCharset utf-8 .html

# All I can say is, we've got more file suffixes than anybody. I mean anybody.
# Every two-bit IF system has its own.

# I think I've listed every non-text file format in the archive, as of
# December 1998. (Updated a bit in August 2003.)
# If you find a file downloading as text/plain that ought
# to be something else, please send mail to <tech .at. ifarchive .dot. org>.

# (This list doesn't include standard mappings, like .gif, which are
# configured in Apache automatically.)

AddType application/octet-stream LHA lha SHK shk SDK sdk arc arj zoo lzh lzc
AddType application/octet-stream acd acx adl agx
AddType application/octet-stream bas bat bsc bxy cbm cfg chm clr cmd con
AddType application/octet-stream d$$ da1 da2 da3 da4 da5 da6 da7 da8 dat
AddType application/octet-stream d64 ddc dif dms dsk elf emb gb gmp hex
AddType application/octet-stream ico img jac map mg1 opt pat patch pif pkg
AddType application/octet-stream rad rpm run sea sfx sna z80
AddType application/octet-stream rs0 rs1 rs2 rs3 rs4 rs5 rs6 rs7 rs8 rs9
AddType application/octet-stream spk spark td0 tos ttl uhs voc
AddType application/octet-stream taf dll DLL x32
AddType application/octet-stream prg PRG

# Mark a few types which are common enough to be worth it.
# I'm going with application/x-zmachine for Z-code game files.
AddType application/x-zmachine z1 z2 z3 z4 z5 z6 z7 z8
AddType application/x-zmachine Z1 Z2 Z3 Z4 Z5 Z6 Z7 Z8
# And, amazingly, TADS seems to be the only system which uses .gam.
AddType application/x-tads gam GAM
AddType application/x-t3vm-image t3 T3
# I'm writing this file, I can include the formats I invented.
AddType application/x-glulx ulx ULX
AddType application/x-blorb blb BLB blorb BLORB
AddType application/x-blorb;profile="zcode" zlb zblorb
AddType application/x-blorb;profile="glulx" glb gblorb

# Mark a few text formats which might be served as something else.
AddType text/plain me t src doc

# The up-to-date type for RSS.
AddType application/rss+xml rss

# Miscellaneous
AddType application/octet-stream exe EXE
AddType application/zip zip ZIP
AddType application/octet-stream prc pdb
AddType application/pdf pdf PDF
AddType application/x-newton-compatible-pkg pkg
AddType application/mac-binhex40 hqx HQX
AddType application/octet-stream cpt CPT
AddType application/x-stuffit sit SIT
AddEncoding x-gzip tgz
AddType application/x-tar tgz
AddType application/java class
AddType audio/x-mp3 mp3
AddType image/png png
AddType image/x-pcx pcx
AddType application/x-bittorrent torrent

# I originally made the default type application/octet-stream, and tried
# to list every suffix used on a text file. Then I gave up.

#DefaultType application/octet-stream
#
#AddType text/plain txt text 1st readme announce from doc letters lst list me
#AddType text/plain c h el inf t pas perl pl notes origin
#AddType text/plain step step1 step2 step3 step4 clue clues sol
#AddType text/plain sup1 sup2 sup3 txt1 txt2 txt3 wlk
#AddType text/plain Readme README FAQ POSTER TXT ME
````

Since this list (154) is far in excess of what's practical to add to command line and obscure enough to not warrant adding to ````gophernicus.h````, I believe that supplying this data in an external file would be warranted.  I chose this format of ````<type> <extension>```` because that's what pygopherd uses.  I'd rather not use pygopherd for a variety of reasons.  I wrote a simple script to convert that mime.conf file into mime.types.

I think I now understand your suggested format.  That would work too and I think may be preferable because it contains more useful information than does mime.types.

**Comment by DavidGriffith** [comment]
By the way, where is the gophertype of "s" defined?

**Comment by hb9kns** [comment]
Now I see your point.
But I think we should add an include to gophernicus.h and not parse this data at runtime, because that might have a noticeable impact on performance.

The FILETYPE format in gophernicus.h is not very nice for input, so we'll need a small script to convert from e.g `0 txt txt1 txt2 md ...` format to the internal `"txt","0","txt1","0","txt2","0","md","0",...` format during compilation.

I don't want to change the internal syntax, because I assume Kim had good reasons for it, and *It Just Works.*

It would then be up to you to convert the mime-type format into gophertype

**Comment by DavidGriffith** [comment]
I can work with that.  The format appears to be a simply pairs of suffix-type repeated with two nulls at the end.  In ````gophernicus.h``` with "sizes and maximums", are the choices based on anything other than someone thinking that powers of two seemed good?

**Comment by fosslinux** [comment]
At any rate, the internal syntax is RFC standard and most gopher clients won't display correctly without the syntax. gophernicus does do some "smart-filling" of some fields sometimes.

**Comment by jamestomasino** [comment]
https://sourceforge.net/p/motsognir/code/HEAD/tree/trunk/extmap.c

Motsognir does this by way of an optional extmap file. You can see it in the documentation, and code above. Good for brainstorming.

Ultimately this is for autogenerated links in directories without gophermaps.

**Comment by fosslinux** [comment]
This was fixed by #57.
